### DISCUSSION TITLE
How to change return type for createFieldExtension

### DISCUSSION THREAD

**Discussion by herecydev** [gatsbyjs/gatsby#27710] [Help] [answered]
Given the following schemaCustomization:

```js
exports.createSchemaCustomization = ({ actions }) => {
  const { createFieldExtension, createTypes } = actions;

  createFieldExtension({
    name: "currencyFormat",
    description: `Add currency formatting options`,
    args: {
      formatCurrency: `Boolean`,
    },
    extend(options, fieldConfig) {
      return {
        ...fieldConfig.args,
        args: {
          formatCurrency: {
            type: `Boolean`,
            description: `Formats price to a currency string`
          },
        },
        async resolve(source, args, context, info) {
          const resolver = fieldConfig.resolve || context.defaultFieldResolver
          const price = await resolver(source, args, context, info)
          return args.formatCurrency ? "some_formatted_string" : price
        },
      }
    },
  })

  createTypes(schema);
};
```

and the following schema:

```js
type SomeType implements Node @dontInfer {
    price: Float! @currencyFormat
  }
```

I want to be able to query the `price` field, so that both the following are possible:

```
{
  price //just returns the float price
}

{
  price(formatCurrency: true) //returns a string
}
```

I feel like this is exactly what Gatsby's @dateFormat directive is already doing when you pass it `(formatString: "YYYY-MM-DD")`. Have I done something wrong here? How can I get the extension to return different types correctly?

**Comment by vladar** [comment, marked as answer]
Hey @herecydev 

Thanks for this question! The field type cannot change after it is defined in the schema. The only way you can do this is by setting a custom scalar type on your field. Gatsby includes the most general type: `JSON` (`Date` is another example but it's always resolves to string). Technically, you can introduce your own custom scalar type via schema customization ([an example](https://github.com/gatsbyjs/gatsby/blob/5611c0f4b1f03662e7f11813d08ee56cd2ca74f5/packages/gatsby/src/schema/types/date.ts#L105-L115)) but this is not documented and I would generally recommend staying away from custom scalar types even in regular GraphQL APIs.

**Reply by herecydev** [reply]
Thanks, I think I understand why the `Date` works then. If you're not recommending using the scalar Types; how would you best solve the problem then?

I've used customResolvers in some places (adding another field with a different return type that can consume the node's data). But there's no reusability, each field is manually added. What I'm after here is a way of the graphql consumer specifying the data should be transformed. It sounds exactly what directives are there for

**Reply by vladar** [reply]
There are a couple of options depending on the situation:

1. Have a `Money` type (with custom resolvers):
```
type Money {
  value: Float
  formatted(currency: CurrencyEnum): String
}
```

2. Have another field `priceFormatted` along with your `price` field.

3. And finally: use a `JSON` type or just a `String` type with a directive. Directives only work when they are compatible with the underlying field type.
